### DISCUSSION TITLE
How to see the ipaddress in the remote logs in place of hostname?

### DISCUSSION THREAD

**Discussion by SrikalaSandu** [syslog-ng/syslog-ng#4338] [Q&A] [answered]
This is my current basic configuration 
```
@version: 3.22
options { 
	flush_lines(0);
	time_reopen(10);
	perm(0664);
	stats_freq(0);
	keep_hostname(no);
	create_dirs(no);
	use_dns(no);
	use_fqdn(no);
};
source s_local { system(); internal(); };
rewrite r_sourceip{
    set('${SOURCEIP}' value(HOST));
};
destination d_syslog_tcp {
syslog("#client_ip" transport("tcp") port(514)); };
log { source(s_local); rewrite(r_sourceip);destination(d_syslog_tcp); };
```
and in my remote logs is seen as this
`month day time 127.0.0.1 systemd[1] Started System Logger Daemon.`

I am unsure why I am seeing a loopback address instead of my client ip.

**Comment by bazsi** [comment]
Because these are local logs emitted by the local systemd.

**Reply by SrikalaSandu** [reply]
hey, I never really got an answer for my question. 
And I'm not quiet sure I understand how it's from my local systemd, my server is a remote machine and I have tried the configuration with various clients.
should I provide more information?

**Comment by bazsi** [comment, marked as answer]
I am not sure I understand your entire topology. You have posted a client
config but can you elaborate where you would like to see what IP address?
You mean that the outgoing message should contain the outgoing IP?

On the config you have posted, the systemd  message is probably coming from
the local source, which means that SOURCEIP is 127.0.0.1.

Then you set HOST to $SOURCEIP that causes a hostname field to be set to
127.0.0.1. The destination will then use this value in the outgoing
message, in place of the hostname.

If you control the server, you could override the incoming hostname there.
This is what keep-hostname(no) and use-dns(no) would do, if you were
setting those on the receiving end.

If you don't control the server or if it's not a syslog-ng instance and
your alternative does not support overriding the hostname, you would need
to set HOST to the outgoing IP, instead of $SOURCEIP, as the latter will
contain 127.0.0.1

syslog-ng doesn't have a built-in macro for querying the outgoing IP
address at the time of sending. You need to set HOST before sending it out
and it is automatically set at syslog-ng reception time (to the name of the
host in case the message is local).

You could use a string literal in your config when you set HOST explicitly,
or use an environment variable and backtick substitution to get the actual
IP address.

**Reply by SrikalaSandu** [reply]
I apologize for the delayed reply.
Thank you for explaining my mistakes.

yes, my server has rsyslog and I have no control over it.

unfortunately setting the host as the ip is not an option for me. in my case there are many machines and setting it for all would be painful. 

can you please elaborate on this ?

> You could use a string literal in your config when you set HOST explicitly,
or use an environment variable and backtick substitution to get the actual
IP address.

### I will go into more detail about what I have done:

this the configuration which has worked for my previous client machine which displays my client's ip address 
```
@version: 3.3.4

options { 
	flush_lines(0);
	time_reopen(10);
	perm(0664);
	stats_freq(0);
	keep_hostname(yes);
	create_dirs(no);
	use_dns(no);
	use_fqdn(no);
};

source s_src { file(/proc/kmsg); unix-stream(/dev/log);internal(); };

rewrite r_rewrite_set{set('', value(HOST));};

destination d_abc { file(/var/log/abc); };
destination 'd_destip' { udp('destip'); };

log { source(s_src); destination(d_facility); destination('d_destip');  rewrite(r_rewrite_set); };
```
### output:

`month day time **client's ip**  syslog-ng[2036]: syslog-ng starting up; version='3.3.4'`

-  This is how I want the output to look like

### The configuration on my recent machine:

I have tried integrating the same configuration onto my recent machine which has updated debian packages etc and only version:3.22 and up works. 
and for 3.22 I need to add **_system();_** parameter in the **_source s_src{}_** 

and this is the recent machine's output with errors: Also would like to mention my hostname is localhost. 
```
month day time localhost syslog-ng[ID]: syslog-ng shutting down; version='3.28.1'
month day time localhost syslog-ng[ID]: syslog-ng starting up; version='3.28.1'
month day time localhost systemd[1]: Stopping System Logger Daemon...
month day time localhost systemd[1]: syslog-ng.service: Succeeded.
month day time localhost systemd[1]: Stopped System Logger Daemon.
month day time localhost systemd[1]: syslog-ng.service: Consumed 21.661s CPU time.
month day time localhost systemd[1]: Starting System Logger Daemon...
month day time localhost syslog-ng[ID]: [ISODATE] Using /dev/log Unix socket with systemd is not possible. Changing to systemd-syslog source, which supports socket activation.;
month day time localhost syslog-ng[ID]: [ISODATE] WARNING: With use-dns(no), dns-cache() will be forced to 'no' too!;
month day time localhost syslog-ng[ID]: [ISODATE] Failed to acquire /run/systemd/journal/syslog socket, disabling systemd-syslog source;
```

**Comment by bazsi** [comment]
On Thu, Feb 23, 2023, 09:28 SrikalaSandu ***@***.***> wrote:

> I apologize for the delayed reply.
>
> this the configuration which has worked for my previous client machine
> which displays my client's ip address
>
> @version: 3.3.4
>
> options {
> 	flush_lines(0);
> 	time_reopen(10);
> 	perm(0664);
> 	stats_freq(0);
> 	keep_hostname(yes);
> 	create_dirs(no);
> 	use_dns(no);
> 	use_fqdn(no);
> };
>
> source s_src { file(/proc/kmsg); unix-stream(/dev/log);internal(); };
>
> rewrite r_rewrite_set{set('', value(HOST));};
>
> destination d_abc { file(/var/log/abc); };
> destination 'd_destip' { udp('destip'); };
>
> log { source(s_src);
>
>
 destination(d_facility);
>
>
destination('d_destip');
>
>
rewrite(r_rewrite_set); };
>
> output:
>
> This is how I want the output to look like
>
> month day time **client's ip** syslog-ng[2036]: syslog-ng starting up;
> version='3.3.4'
>
> I have tried integrating this into my recent machine which has updated
> debian packages etc and only version:3.22 and up works.
> and for 3.22 I need to add *system();* parameter in the *source s_src{}*
>
> and this is the recent machine's output with errors:
>
> month day time localhost syslog-ng[ID]: syslog-ng shutting down; version='3.28.1'
> month day time localhost syslog-ng[ID]: syslog-ng starting up; version='3.28.1'
> month day time localhost systemd[1]: Stopping System Logger Daemon...
> month day time localhost systemd[1]: syslog-ng.service: Succeeded.
> month day time localhost systemd[1]: Stopped System Logger Daemon.
> month day time localhost systemd[1]: syslog-ng.service: Consumed 21.661s CPU time.
> month day time localhost systemd[1]: Starting System Logger Daemon...
> month day time localhost syslog-ng[ID]: [ISODATE] Using /dev/log Unix socket with systemd is not possible. Changing to systemd-syslog source, which supports socket activation.;
> month day time localhost syslog-ng[ID]: [ISODATE] WARNING: With use-dns(no), dns-cache() will be forced to 'no' too!;
> month day time localhost syslog-ng[ID]: [ISODATE] Failed to acquire /run/systemd/journal/syslog socket, disabling systemd-syslog source;
>
> my hostname is localhost.
>
>

Is this log visible in the file /var/log/abc or on the target system where
you are sending the message via UDP?

It might matter, as the file is written on the local host where the IP
address does not play a role. If it's fine by the server receiving this
message, an empty host name will cause the receiver to resolve the sending
IP address and place it in the HOST field.

Your rewrite rule doesn't seem to do anything as it's located after the
destination.

Also d_abc is not referenced while d_facility is, but that's not present in
the sample config above
